vms/evm/uptimetracker/validator.go

```diff
@@ -0,0 +1,329 @@
+}
+
+// GetValidationID returns the validation ID for the given nodeID
+func (u *UptimeTracker) GetValidationID(nodeID ids.NodeID) (ids.ID, bool) {
+	vID, ok := u.state.index[nodeID]
+	if !ok {
+		return ids.ID{}, false
+	}
+	return vID, true
+}
+
+// GetValidator returns the validator data for the given validationID
+func (u *UptimeTracker) GetValidator(vID ids.ID) (Validator, bool) {
+	data, ok := u.state.data[vID]
+	if !ok {
+		return Validator{}, false
+	}
+	return Validator{
+		ValidationID:   data.validationID,
+		NodeID:         data.NodeID,
+		StartTimestamp: data.StartTime,
+		IsActive:       data.IsActive,
+		Weight:         data.Weight,
+		IsL1Validator:  data.IsL1Validator,
+	}, true
+}
+
+// GetValidators returns all validators in the state.
+func (u *UptimeTracker) GetValidators() []Validator {
+	validators := make([]Validator, 0, len(u.state.data))
+	for _, vdr := range u.state.data {
+		validators = append(validators, Validator{
+			ValidationID:   vdr.validationID,
+			NodeID:         vdr.NodeID,
+			Weight:         vdr.Weight,
+			StartTimestamp: vdr.StartTime,
+			IsActive:       vdr.IsActive,
+			IsL1Validator:  vdr.IsL1Validator,
+		})
+	}
+	return validators
+}
+
+// GetValidatorAndUptime returns the calculated uptime of the validator specified by [validationID]
+// and the last updated time.
+// GetValidatorAndUptime holds the lock while performing the operation and can be called concurrently.
+func (u *UptimeTracker) GetValidatorAndUptime(validationID ids.ID, lock sync.Locker) (Validator, time.Duration, time.Time, error) {
+	lock.Lock()
+	defer lock.Unlock()
+
+	vdr, ok := u.GetValidator(validationID)
+	if !ok {
+		return Validator{}, 0, time.Time{}, fmt.Errorf("failed to get validator %s", validationID)
+	}
+
+	uptime, lastUpdated, err := u.manager.CalculateUptime(vdr.NodeID)
+	if err != nil {
+		return Validator{}, 0, time.Time{}, fmt.Errorf("failed to calculate uptime for validator %s: %w", validationID, err)
+	}
+
+	return vdr, uptime, lastUpdated, nil
+}
+
+// OnValidatorAdded is called when a validator is added.
+// If the node is inactive, it will be paused.
+func (u *UptimeTracker) onValidatorAdded(_ ids.ID, nodeID ids.NodeID, _ uint64, isActive bool) {
+	if !isActive {
+		err := u.pause(nodeID)
+		if err != nil {
+			log.Error("failed to handle added validator %s: %s", nodeID, err)
+		}
+	}
+}
+
+// OnValidatorRemoved is called when a validator is removed.
+// If the node is already paused, it will be resumed.
+func (u *UptimeTracker) onValidatorRemoved(_ ids.ID, nodeID ids.NodeID) {
+	if u.isPaused(nodeID) {
+		err := u.resume(nodeID)
+		if err != nil {
+			log.Error("failed to handle validator removed %s: %s", nodeID, err)
+		}
+	}
+}
+
+// OnValidatorStatusUpdated is called when the status of a validator is updated.
+// If the node is active, it will be resumed. If the node is inactive, it will be paused.
+func (u *UptimeTracker) onValidatorStatusUpdated(_ ids.ID, nodeID ids.NodeID, isActive bool) {
```

Comment:
This looks like vestigial code from callback hell we previously had - could we just in-line usage of it? Generally I dislike these types of functions when they wrap trivial logic + are used in only one place.